### ISSUE TITLE
Drop Kyber support

### ISSUE THREAD

**Issue by dstebila** [open-quantum-safe/liboqs#1989]
With ML-KEM support now available in liboqs and also in other TLS implementations (https://github.com/open-quantum-safe/oqs-provider/issues/561#issuecomment-2469455998), it's been suggested in https://github.com/open-quantum-safe/oqs-provider/issues/561#issuecomment-2469793743 that we drop Kyber support from liboqs and oqs-provider.  Thoughts?

**Comment by dstebila** [comment]
In today's OQS status call, I suggested that we keep Kyber in the forthcoming liboqs 0.12.0 release, and then drop Kyber in 0.13.0. We would include a note in the 0.12.0 release notes warning that this is the last release with Kyber.

We also discussed the impact this has on oqs-provider, and specifically re-assigning TLS hybrid code-points to the reserved range (https://github.com/open-quantum-safe/oqs-provider/issues/561).  It seems convenient to do that re-assignment at the same time as dropping Kyber support.

**Comment by loganaden** [comment]
I would suggest against doing that. There are commercial products that need to keep compatibility with Kyber768 while at the same time offering ML-KEM 768 support. Clients who are still on Kyber 768 need to be supported.

**Comment by baentsch** [comment]
> There are commercial products that need to keep compatibility with Kyber768 while at the same time offering ML-KEM 768 support. 

What products are those? Why is this relevant for `liboqs`? 

There is a disclaimer recommending against commercial use of OQS -- partially exactly for the reason for OQS not having (or rather, not being able) to support outdated software for a long time -- and even for free: _Commercial_ products may offer long term or legacy support. There's whole companies only alive because of (people paying for) legacy software but I don't see why volunteers should do such work instead of doing new things.

> Clients who are still on Kyber 512/768/1024 need to be supported.

What clients are those? If those clients use OQS, can they be motivated to help OQS maintaining this algorithm family?

**Comment by koraa** [comment]
Hi @baentsch, 

@loganaden is helping us with the migration to ML-KEM in [Rosenpass](https://github.com/rosenpass/rosenpass) (PQ security for WireGuard and a generic mutually authenticated kem-based key exchange).

Our interest is mainly not giving our users a frustrating experience when they update their Rosenpass implementations; very long term support of Kyber is not a goal and we think people should migrate to ML-KEM eventually.

We are an open source project too but I am happy to chat about figuring out how support Kyber for longer.

**Comment by koraa** [comment]
And to be a bit more specific here: We don't have commercial clients right now for whom migrating would be a major issue. We have users who we would like to avoid alienating.

**Comment by dstebila** [comment]
Hi Karolin! Thanks for the feedback. Do you have a sense of what your timeline is for removal of Kyber support?

**Comment by koraa** [comment]
@dstebila Thank you for getting back! We would be targetting a transition period of a couple of years. Long enough for us to:

1. Make a release with the new version
2. Have distributions adopt the new release
3. Wait for a year or more to give everyone ample time to upgrade

It would be possible to shorten the transition period by relying on two versions of liboqs, one outdated, deprecated with Kyber.

**Comment by baentsch** [comment]
Hi @koraa -- great to get in touch and thanks for the background information. Always good to know actual users of `liboqs` (thanks for the link-up @loganaden !). Also, funnily enough you also seem to have received funding from the great NLNet team which also allowed me to buy some bread while working on [oqs-provider](https://github.com/open-quantum-safe/oqs-provider) a few years back.

This statement is good

> very long term support of Kyber is not a goal

but this gives me the creeps:

> We would be targetting a transition period of a couple of years

So, allow me to understand how you can provide support over such long periods of time?

Regarding the procedure: Did you integrate `liboqs` via a specific release/tag? If so, why can't Rosenpass simply retain that and let `liboqs` move on? No-one will notice Kyber being removed from `liboqs` "main" and new releases as Rosenpass will keep using that tag/release, until you do an ML-KEM-based release of Rosenpass (then fixed to some more current `liboqs` tag/release), right? Or do you have maintenance expectations towards `liboqs` (particularly old releases)? If so, which ones?

Finally, may I ask about the reason to not fully integrate PQC into wireguard itself? If I don't misunderstand the purpose of Rosenpass

> Rosenpass is used to create post-quantum-secure VPNs. 

wouldn't this be a more desirable medium-to-long term goal over a "couple of years transition period" (and continued maintenance) for Rosenpass? Also, I'd be curious to understand (been a researcher myself a long time ago) what's the conceptual advantage of Rosenpass compared to say [a generic PQC-enhanced software stack like OpenVPN](https://github.com/open-quantum-safe/oqs-demos/tree/main/openvpn)?

**Comment by dstebila** [comment]
> Regarding the procedure: Did you integrate `liboqs` via a specific release/tag? If so, why can't Rosenpass simply retain that and let `liboqs` move on? No-one will notice Kyber being removed from `liboqs` "main" and new releases as Rosenpass will keep using that tag/release, until you do an ML-KEM-based release of Rosenpass (then fixed to some more current `liboqs` tag/release), right? Or do you have maintenance expectations towards `liboqs` (particularly old releases)? If so, which ones?

This seems suboptimal.  While one could get Kyber from an old tag of liboqs, presumably it would be preferably to be using ML-KEM from more recent builds of liboqs as that implementation evolves/improves.  I would imagine that using two copies of liboqs within the same binary is quite tricky.

**Comment by baentsch** [comment]
> I would imagine that using two copies of liboqs within the same binary is quite tricky.

Why would you (well, rather @koraa) do that? There would be one release (binary?) of Rosenpass (v0.x) with Kyber (and the corresponding `liboqs` version) and another/separate/future one (v0.x++) with ML-KEM (exclusively) --with a separate `liboqs` version--, no? Or do you indeed want to run both algorithms at the same time?

@koraa: Can you please shed some more light on how you use (and intend to use) `liboqs`? For example, is your app statically or dynamically linked to `liboqs`? Also, do you use [liboqs-rust](https://github.com/open-quantum-safe/liboqs-rust) or are you bypassing that language wrapper? If the latter, what requirements do you have towards that project?

**Comment by dstebila** [comment]
> > I would imagine that using two copies of liboqs within the same binary is quite tricky.
> 
> Why would you (well, rather [@koraa](https://github.com/koraa)) do that? There would be one release (binary?) of Rosenpass (v0.x) with Kyber (and the corresponding `liboqs` version) and another/separate/future one (v0.x++) with ML-KEM (exclusively) --with a separate `liboqs` version--, no? Or do you indeed want to run both algorithms at the same time?

Interoperability? A server would need to support old clients that are using Kyber and new clients using ML-KEM, and hence would need to have code for both algorithms simultaneously. Presumably the same for new clients that still want to be able to talk to old servers.

**Comment by koraa** [comment]
Hi @baentsch, hi @dstebila,

thank you for your responses and your insight.


> This statement is good
>
>> very long term support of Kyber is not a goal
>
> but this gives me the creeps:
>
>> We would be targetting a transition period of a couple of years

I am glad you are happy about our plans regarding Kyber, and I am sorry our intentions about deprecation strategies have affected you in this way.

> So, allow me to understand how you can provide support over such long periods of time?

The short answer is: It is not clear that this is a goal that can be accomplished, but we try as best as we can. In the name of usable security, we are trying hard to avoid friction for our users.

> Regarding the procedure: Did you integrate liboqs via a specific release/tag? 

We are directly relying on oqs-sys:

```toml
oqs-sys = { version = "0.9.1", default-features = false, features = [
    'classic_mceliece',
    'kyber',
]}
```

> If so, why can't Rosenpass simply retain that and let liboqs move on? No-one will notice Kyber being removed from liboqs "main" and new releases as Rosenpass will keep using that tag/release, until you do an ML-KEM-based release of Rosenpass (then fixed to some more current liboqs tag/release), right?

It is exactly as @dstebila said: 

> This seems suboptimal. While one could get Kyber from an old tag of liboqs, presumably it would be preferably to be using ML-KEM from more recent builds of liboqs as that implementation evolves/improves. I would imagine that using two copies of liboqs within the same binary is quite tricky.

Having two copies of liboqs can probably be done, but yes -- it would be tricky and brittle.

And @dstebila's second point covers our perspective regarding your follow-up question:

> Why would you (well, rather @koraa) do that? There would be one release (binary?) of Rosenpass (v0.x) with Kyber (and the corresponding liboqs version) and another/separate/future one (v0.x++) with ML-KEM (exclusively) --with a separate liboqs version--, no? Or do you indeed want to run both algorithms at the same time?

> Interoperability? A server would need to support old clients that are using Kyber and new clients using ML-KEM, and hence would need to have code for both algorithms simultaneously. Presumably the same for new clients that still want to be able to talk to old servers.

In a nutshell: Imagine you are using Arch Linux on your client and Debian on your server, and you run a system update on both; Debian will update to the newest Rosenpass version after years; Arch will be quick. By using a hard deprecation strategy, we are giving this particular user a hard time because they now have to install a package on either system from outside the normal package users.

I do not think Debian currently integrates Rosenpass, but Arch Linux does. There are a variety of reasons why versions might be out of sync on client and server.

> Or do you have maintenance expectations towards liboqs (particularly old releases)? If so, which ones?

This question comes as a bit of a surprise for me. There is no expectation we place on liboqs other than the usual expectations of collegial cooperation between open-source projects working on the shared goal of a post-quantum transition. We certainly do not have the expectation of an explicit or implicit service-level agreement.

I do like to believe that we maintain close, mutually beneficial relationships with other members of the real-world cryptography community, and I certainly believe working together to come up with the best deprecation strategies for something like Kyber is in the interest of us all.

Nothing is written in stone, and we do not have all the wisdom. We haven't depracted intermediate versions of newly standardized cryptographic primitives before either, and so I think we are in this together to figure out the best path forward.

That said, Rosenpass is one of a probably small number of open-source post-quantum projects using liboqs that do have quite a few real-world users, so I think we might be a useful case study for you.

> Finally, may I ask about the reason to not fully integrate PQC into wireguard itself? If I don't misunderstand the purpose of Rosenpass

This question seems a bit out of scope for this particular venue. I am happy to jump on an email thread and discuss the design of Rosenpass if you are interested.

My e-mail address is karo@rosenpass.eu.

> > Rosenpass is used to create post-quantum-secure VPNs.
> 
> wouldn't this be a more desirable medium-to-long term goal over a "couple of years transition period" (and continued maintenance) for Rosenpass

Could you rephrase that question? I do not think I understand.

> Also, I'd be curious to understand (been a researcher myself a long time ago) what's the conceptual advantage of Rosenpass compared to say a generic PQC-enhanced software stack like OpenVPN?

I think this query too might better be answered in another channel. I am happy to discuss, and we could collaborate to write a blog post, so the answers may be beneficial to all.

Thank you for the stimulating discussion to both of you!

**Comment by dstebila** [comment]
> Nothing is written in stone, and we do not have all the wisdom. We haven't depracted intermediate versions of newly standardized cryptographic primitives before either, and so I think we are in this together to figure out the best path forward.
> 
> That said, Rosenpass is one of a probably small number of open-source post-quantum projects using liboqs that do have quite a few real-world users, so I think we might be a useful case study for you.

Indeed, I think the information you've given us is very useful. We need to be more proactive in trying to understand who is using liboqs and what they want from the project, and your perspective is really helpful.

**Comment by baentsch** [comment]
> I do like to believe that we maintain close, mutually beneficial relationships with other members of the real-world cryptography community, and I certainly believe working together to come up with the best deprecation strategies for something like Kyber is in the interest of us all.

Agreed. Well, a year ago I had agreed more and wouldn't have asked a single question but just supported you(r app). Now, however, there's corporate entities that have taken control of OQS, not providing incremental funding or staffing but just adding demands and creating expectations the few people contributing or maintaining have to fulfil (or, rather juggle with sensible expectations like yours), see e.g., [this](https://pqca.org/blog/2024/pqca-and-nists-post-quantum-cryptography-standards/): 

> To address research, the [PQCA project lifecycle](https://pqca.github.io/TAC/Processes/Project_Lifecycle.html) policy was explicitly designed to support research implementations as well as production code, and we have both research tracks and production tracks for projects.  This dual-track approach will let OQS and its contributors to be able to continue to be leaders in research as well as move towards production code

(which actually already is a toned-down version as per https://github.com/PQCA/TAC/issues/42#issuecomment-2497063996 : And that shows another issue: The technical contribution activity of the formerly two most active contributors, @dstebila and myself have come to a crawl trying to "reign in" or try to fulfil an increase in commerce-oriented marketing activity /expectation setting not backed by a commensurate increase in technical contributions/activity -- the opposite actually, as per the above).

Back to the issue: I agree that retaining support for all algorithms that OQS ever supported is ideal for consumers, no doubt. At the same time, the [PQCA-driven](https://github.com/PQCA/TAC/issues/44) "productization" pressure creates process demands like [security issues handling](https://github.com/open-quantum-safe/tsc/pull/124) that's not just adding work and responsibility -- but also multiplies this by the number of algorithms supported.

Therefore, apologies to @koraa for the questions above: They've all been borne of the wish to keep OQS maintainable for a longer period of time and any support by the community (incl. users) to do so would be gratefully welcome. The other questions I'll take to email as per your suggestion.

**Comment by koraa** [comment]
> Indeed, I think the information you've given us is very useful. We need to be more proactive in trying to understand who is using liboqs and what they want from the project, and your perspective is really helpful.

@dstebila Please consider us as having issued a standing invitation to collaborate on discussions, provide input, or collaborate on gathering empirical input regarding this topic or similar topics of how to maintain or governs PQC projects in practice. We are also in the process of learning :)

> Agreed. Well, a year ago I had agreed more and wouldn't have asked a single question but just supported you(r app). Now, however, there's corporate entities that have taken control of OQS, not providing incremental funding or staffing but just adding demands and creating expectations the few people contributing or maintaining have to fulfil (or, rather juggle with sensible expectations like yours), see e.g., [this](https://pqca.org/blog/2024/pqca-and-nists-post-quantum-cryptography-standards/):

I hear your frustration. There is no comprehensive answer I can offer other than a general critique of capitalism.

Open-Source projects such as OQS or Rosenpass are becoming relevant in the real world; one aspect of this is interest from companies, and companies are always interested in getting the most and giving the least. It's not a symbiotic relationship; it's a parasitic one.

When I began Rosenpass and when I decided to utilize a permissive FOSS license, I did so expecting and knowing that companies would exploit my work and that they would not give back appropriately. I can name concrete instances where that did happen.

The only advice I can give is that freedom from such exploitation is not found in opposition to it; in fact, the act of opposing someone or something is always also an expression of the power those opposed have over you. This is part of the reason why I tend to ignore the exploitation unless it's actively harmful, *and* I have the means to do something against it.

Instead, I use the resources at our disposal as best I can; I manage some cash flows, make sure contributors at Rosenpass are getting paid, and I use our impact on companies to strengthen my negotiation position.

I can not act outside of a capitalist system, I can only try to serve my group as best as I can and represent our interests.

> Back to the issue: I agree that retaining support for all algorithms that OQS ever supported is ideal for consumers, no doubt. At the same time, the [PQCA-driven](https://github.com/PQCA/TAC/issues/44) "productization" pressure creates process demands like [security issues handling](https://github.com/open-quantum-safe/tsc/pull/124) that's not just adding work and responsibility -- but also multiplies this by the number of algorithms supported.

This is an issue your community probably needs to discuss internally. I would just offer the following standpoint:

I would think actively about the customer; both the commercial customer who should pay their share and the community customer, who you are trying to serve. I would focus on the community customer and ignore the commercial one.

It is possible to use the financial interests of companies to your advantage and to cross finance work for the community, but it is extra work and not necessarily the sort of work you want to do.

> Therefore, apologies to [@koraa](https://github.com/koraa) for the questions above: They've all been borne of the wish to keep OQS maintainable for a longer period of time and any support by the community (incl. users) to do so would be gratefully welcome. The other questions I'll take to email as per your suggestion.

No offense taken :)
These are important conversations to be had!
Thanks for your E-Mail, I'll respond shortly.

**Comment by SWilson4** [comment]
In light of the discussion in this issue, we will not be removing Kyber in the upcoming 0.13.0 release, despite our expressed intent to do so in the [0.12.0 release notes](https://github.com/open-quantum-safe/liboqs/releases/tag/0.12.0).

**Comment by koraa** [comment]
@SWilson4 That is very kind. Thank you! :)

**Comment by dstebila** [comment]
[PQClean will start winding down](https://github.com/PQClean/PQClean/issues/604#issuecomment-3724342593) with a plan to switch to read-only in about 6 months time. liboqs has been relying on PQClean for the Kyber aarch64 optimized implementation, so we are discussing in https://github.com/open-quantum-safe/liboqs/issues/2341 what impact this will have on liboqs. One option raised is to drop Kyber, which we discussed last year on this issue. 

I wanted to check in again to see if any of the people who commented on this issue are still relying on liboqs for Kyber.

**Comment by koraa** [comment]
@dstebila We are still using it and we will be offering kyber-support as a backwards-compatibility feature for the forseeable future.

**Comment by dstebila** [comment]
> [@dstebila](https://github.com/dstebila) We are still using it and we will be offering kyber-support as a backwards-compatibility feature for the forseeable future.

Thanks for the info, Karolin. If we replace the aarch64 optimized Kyber implementation with the C reference implementation, would that cause you significant problems?

**Comment by koraa** [comment]
> > [@dstebila](https://github.com/dstebila) We are still using it and we will be offering kyber-support as a backwards-compatibility feature for the forseeable future.
> 
> Thanks for the info, Karolin. If we replace the aarch64 optimized Kyber implementation with the C reference implementation, would that cause you significant problems?

I don't know TBH; it might cause a performance regression, but especially as a backwards compatibility feature this won't bee too much of a concern.

**Comment by ghen2** [comment]
I assume the need for a supported Kyber implementation is primarily on the server-side, to support outdated (but PQC-capable) clients?

While those Kyber clients may run on Aarch64, they are by definition outdated, so not using the latest liboqs. :)
(up-to-date clients support ML-KEM now)

**Comment by dstebila** [comment]
Can you help us come up with a timeline for decreasing and then dropping Kyber support?  For example, drop the aarch64 optimized implementation in say 4 months due to the PQClean wind-down, and then remove Kyber 1 year from now? I think we don't want an indefinite support cycle for a deprecated algorithm, and having a set timeline would help us decide how much resources we want to devote to Kyber.

**Comment by ghen2** [comment]
I guess it mostly depends on the client base that projects like Rosenpass (@koora) want to continue to support.
Kyber-capable but ML-KEM-incapable clients are "recent but not the latest", a category of clients that should be doable to get upgraded?   So 1 more year seems already quite long.

@bwesterb Cloudflare also still supports X25519Kyber768 (`0x6399`), how much usage are you still seeing there?

**Comment by bwesterb** [comment]
2%

**Comment by ghen2** [comment]
That's still a lot.  Are those real browsers, or mostly bots/crawlers?

**Comment by bwesterb** [comment]
Both.

**Comment by dstebila** [comment]
> I guess it mostly depends on the client base that projects like Rosenpass ([@Koora](https://github.com/Koora)) want to continue to support. Kyber-capable but ML-KEM-incapable clients are "recent but not the latest", a category of clients that should be doable to get upgraded? So 1 more year seems already quite long.

@Koora do you have any thoughts on a timeline that would be appropriate for winding down Kyber support in your application?

**Comment by koraa** [comment]
@dstebila Right now we do not have any concrete plans to remove Kyber/Round 3 support any time soon. I think a deprecation period of between 3 and 5 years should give everyone the opportunity to migrate though.

**Comment by bhess** [comment]
A quick follow-up, as this was discussed at the OQS TSC meeting on [2026-03-24](https://github.com/open-quantum-safe/tsc/blob/main/meetings/2026-03-24/minutes.md#timeline-for-discontinuing-kyber-support).

Given that Kyber has been deprecated and is no longer supported upstream (with PQClean winding down, see #2341, and the recommendation to move from pq-crystals/kyber to mlkem-native), the TSC discussed whether continued long‑term maintenance by the liboqs core team is sustainable.

The model that was discussed was to keep the algorithm, but with an external interested party taking on maintainership.

@koraa: would you and the Rosenpass team be interested in maintaining Kyber in liboqs for the duration of your support window? Concretely, this would involve a CODEOWNERS entry, handling issues and security reports, and ensuring the code continues to build and test on supported platforms. Happy to discuss the specifics.

**Comment by loganaden** [comment]
@koraa I can do it ?

**Comment by koraa** [comment]
@loganaden Absolutely! The contribution would be very welcome. We should regularly in touch every couple of months/weeks though if you provide this service for our project. I am getting in touch by matrix.

**Comment by dstebila** [comment]
I've created https://github.com/open-quantum-safe/liboqs/pull/2421 and https://github.com/open-quantum-safe/tsc/pull/269/changes/0a5bfc01831d67f603d44e70aa463597233577c0 to add @loganaden as a CODEOWNER for Kyber. Once those lands, I would close this issue as won't-fix, since we have a resolution for a while.
